sql: add support for DECLARE, FETCH, CLOSE CURSOR using internal executors

jordanlewis opened the pull request:
Touches #41412.

This PR implements the SQL CURSOR methods DECLARE, FETCH and CLOSE using streaming internal executors.

Here's the Postgres docs on DECLARE: https://www.postgresql.org/docs/current/sql-declare.html

Things it implements:

- DECLARE for forward-only cursors
- FETCH forward, relative, and absolute variants for forward-only cursors
- CLOSE a cursor
- pg_catalog.pg_cursors

Things it's missing:
- BINARY CURSOR support, which returns data in the Postgres binary format
- MOVE support, which allows advancing the cursor without returning any rows. This should be quite similar to FETCH.
- WITH HOLD support (allows keeping a cursor open for longer than a transaction by writing its results into a buffer)
- Scrollable cursor (reverse fetch) support, we'd have to implement this also by savings results into a buffer.
- FOR UPDATE support
- Respect for `SAVEPOINT`. Cursor definitions will not disappear if rolled back to a savepoint before they were created. But we also don't do this right for settings (see #69396) so I don't think this is a major problem.

jordanlewis committed:
sql: add no-op DECLARE, FETCH, and CLOSE support

This commit adds parser, AST, and no-op execution support for the
DECLARE, FETCH, and CLOSE SQL cursor statements.

jordanlewis committed:
sql: add implementation of DECLARE, FETCH and CLOSE

So far, only the simplest form of DECLARE is supported: forward-only
cursors that are not able to be held open past the end of a transaction.

And FETCH only supports the simple form that returns the next n rows.

Release note (sql change): add support for DECLARE, FETCH and CLOSE
commands for creating, using and deleting SQL cursors.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74006)
<!-- Reviewable:end -->

jordanlewis commented:
Thank you to @otan for inspiring me to take a look at this on flex friday!

jordanlewis committed:
sql: move cursor state to connExecutor.extraTxnState

Move the map of session cursors to extraTxnState, which is where similar
state exists, like the list of currently active prepared statements.

Release note: None

jordanlewis committed:
sql: implement pg_catalog.pg_cursors

Add open SQL cursors to pg_catalog.pg_cursors.

Release note (sql change): SQL cursors now appear in
pg_catalog.pg_cursors.

jordanlewis commented:
Alright, I think this is ready for a look.

ajwerner commented:
My first instinct is that this is going to have bad interactions with schema changes in the transaction. That is to say, it won't see them. That's pretty gross. I wonder what happens when you create cursors over various pg_catalog things and then modify them while holding the cursor open. I'm not even sure what the right thing is.

What sort of isolation are cursors supposed to have? I think because you're using the transaction object itself, it's going to be weird. Maybe it should be at some fixed sequence in the transaction?

jordanlewis commented:
> My first instinct is that this is going to have bad interactions with schema changes in the transaction. That is to say, it won't see them. That's pretty gross. I wonder what happens when you create cursors over various pg_catalog things and then modify them while holding the cursor open. I'm not even sure what the right thing is.

This is a good point. I will test to see how it works in PG.

> What sort of isolation are cursors supposed to have? I think because you're using the transaction object itself, it's going to be weird. Maybe it should be at some fixed sequence in the transaction?

Cursors are supposed to see a snapshot at the point of time that they were created in the transaction. That is, they're not supposed to be able to read other writes from the same transaction that happen after they were created.

This behavior is implemented by the current PR, though to be honest, I don't know why it works. I assume when we make an InternalExecutor, we create a sequence point, or something? There's a logic test for it.

ajwerner commented:
> This behavior is implemented by the current PR, though to be honest, I don't know why it works. I assume when we make an InternalExecutor, we create a sequence point, or something? There's a logic test for it.

It'll work for queries where we construct a leaf transaction if you have flows. The leaf transaction occurs at a snapshot. However, it's really important that when this happens we make sure that we properly allow the flows to finish so that the final leaf state flows back to the root txn before committing. Otherwise you'll subtly permit serializability violations. 
 https://github.com/cockroachdb/cockroach/blob/775c2b0b0302ebf4b0c8c8515e015b793b8585ce/pkg/sql/distsql/server.go#L289-L291

I don't know enough to say how fragile this is.

jordanlewis commented:
> It'll work for queries where we construct a leaf transaction if you have flows. 

Hmm, but the tests in this PR work even without leaf transactions - we don't use those in the `local` logic test test config. So there must be something else going on too, right?

ajwerner commented:
> > It'll work for queries where we construct a leaf transaction if you have flows.
> 
> Hmm, but the tests in this PR work even without leaf transactions - we don't use those in the `local` logic test test config. So there must be something else going on too, right?

It'd be interesting to look at traces. Are all of the rows being fetched by the row fetcher in the first request? What if you add a lot of rows or something like that?

jordanlewis commented:
It doesn't appear to have to do with that. I'll pull KV trace next, it's a bit annoying due to internal executor plumbing.

```
demo@127.0.0.1:26257/defaultdb> create table a (a primary key, b, c) as select g,g+1,g+2 from generate_series(1,20000) g(g);
CREATE TABLE AS
demo@127.0.0.1:26257/defaultdb> begin;
BEGIN
demo@127.0.0.1:26257/defaultdb  OPEN> declare foo cursor for select * from a;
DECLARE
demo@127.0.0.1:26257/defaultdb  OPEN> insert into a values(10000000,1,1);
INSERT 1
demo@127.0.0.1:26257/defaultdb  OPEN> fetch relative 19999 foo;
    a   |   b   |   c
--------+-------+--------
  19999 | 20000 | 20001
(1 row)
demo@127.0.0.1:26257/defaultdb  OPEN> fetch 1 foo;
    a   |   b   |   c
--------+-------+--------
  20000 | 20001 | 20002
(1 row)
demo@127.0.0.1:26257/defaultdb  OPEN> fetch 1 foo;
FETCH

Time: 0ms total (execution 0ms / network 0ms)
```

jordanlewis commented:
Ah, I forgot that we changed the KV protocol to not stop sending until it hit a bytes limit as well, not just a row limit. I think your hypothesis is correct.

yuzefovich reviewed (commented):
> Reviewed 16 of 16 files at r1, 6 of 6 files at r2, 4 of 4 files at r3, 6 of 6 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74006)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jordanlewis, @otan, and @rafiss)

___
*[-- commits, line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-MrUJIsiExFUrz4RDQgU):*
nit: missing release note.
___
*[pkg/sql/sql_cursor.go, line 132 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-MrY45xZ74mjHwSXGu21:-MrY45xaAB1sStKFjtjI:bgydyee) ([raw file](https://github.com/cockroachdb/cockroach/blob/b21c6703a2078f37f07ee4ecc5fe2b6380e0074e/pkg/sql/sql_cursor.go#L132)):*
> ```Go
> 			return true, nil
> 		case tree.FetchRelative:
> 			for i := int64(0); i < f.offset; i++ {
> ```

Relative fetch can also be backward, should we check that `f.offset` is not negative?
___
*[pkg/sql/sql_cursor.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-MrY2slW7Iz032GF9Tc5:-MrY2slW7Iz032GF9Tc6:bdu873t) ([raw file](https://github.com/cockroachdb/cockroach/blob/b21c6703a2078f37f07ee4ecc5fe2b6380e0074e/pkg/sql/sql_cursor.go#L175)):*
> ```Go
> }
> 
> func (s *sqlCursor) Next(ctx context.Context) (bool, error) {
> ```

nit: this probably could be unexported.
___
*[pkg/sql/sql_cursor.go, line 55 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-MrY5bzlEVYKY1oMEpWD:-MrY5bzlEVYKY1oMEpWE:b-kjizb) ([raw file](https://github.com/cockroachdb/cockroach/blob/be7c2905a08c7cc35409d803beab9717c555fb94/pkg/sql/sql_cursor.go#L55)):*
> ```Go
> 	cursor := &sqlCursor{InternalRows: rows}
> 	if err := p.sqlCursors.addCursor(cursorName, cursor); err != nil {
> 		// thread at a time. But let's be diligent and clean up if it somehow does
> ```

nit: the comment is cut off.
___
*[pkg/sql/sql_cursor.go, line 182 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-MrY5nObDb89SCjA9Beh:-MrY5nOcAJMar8a9ZPjQ:b-op9ebt) ([raw file](https://github.com/cockroachdb/cockroach/blob/be7c2905a08c7cc35409d803beab9717c555fb94/pkg/sql/sql_cursor.go#L182)):*
> ```Go
> 	// closeAll closes all cursors in the set.
> 	closeAll()
> 	// closeCursors closes the named cursor, returning an error if that cursor
> ```

nit: `s/closeCursors/closeCursor/`.
___
*[pkg/sql/sql_cursor.go, line 185 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-MrY5s1jA7uWAhbgDica:-MrY5s1jA7uWAhbgDicb:b-wiz8fb) ([raw file](https://github.com/cockroachdb/cockroach/blob/be7c2905a08c7cc35409d803beab9717c555fb94/pkg/sql/sql_cursor.go#L185)):*
> ```Go
> 	// didn't exist in the set.
> 	closeCursor(string) error
> 	// closeCursors returns the named cursor, returning an error if that cursor
> ```

nit: `s/closeCursors/getCursor/`.
___
*[pkg/sql/parser/sql.y, line 5092 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-MrULCly4eWohgHPq354:-MrULClzDOKU8dSK4Lwi:b-2t4azn) ([raw file](https://github.com/cockroachdb/cockroach/blob/10edf738171e7d94bff99a94b14941a15a1dc4f9/pkg/sql/parser/sql.y#L5092)):*
> ```Yacc
> // %SeeAlso: CLOSE, FETCH
> declare_cursor_stmt:
> 	DECLARE cursor_name opt_binary opt_sensitivity opt_scroll CURSOR opt_hold FOR select_stmt
> ```

PG docs say that
```
The key words ASENSITIVE, BINARY, INSENSITIVE, and SCROLL can appear in any order.
```
it looks like we currently don't support that, right? Is the only way to support that is to have all permutations defined explicitly?
___
*[pkg/sql/parser/sql.y, line 5195 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-MrULe5b7Fr8NEbYFpKH:-MrULe5b7Fr8NEbYFpKI:b2zjwlb) ([raw file](https://github.com/cockroachdb/cockroach/blob/10edf738171e7d94bff99a94b14941a15a1dc4f9/pkg/sql/parser/sql.y#L5195)):*
> ```Yacc
>       Name: tree.Name($3),
>       Count: $1.int64(),
> 	  }
> ```

nit: there appear to be tabs for seven options in a row here.
___
*[pkg/sql/sem/tree/cursor.go, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-MrY0CyX09BmlKoqcBHb:-MrY0CyX09BmlKoqcBHc:b-wu1o7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/10edf738171e7d94bff99a94b14941a15a1dc4f9/pkg/sql/sem/tree/cursor.go#L42)):*
> ```Go
> 	}
> 	ctx.WriteString("CURSOR ")
> 	if node.Hold {
> ```

Does it not matter to distinguish between unspecified hold and explicit "without hold"?


<!-- Sent from Reviewable.io -->

jordanlewis committed:
sql: make cursors "insensitive"

This commit causes cursors to be "insensitive", aka not reading writes
that happen after they were declared in a transaction, by fixing the
transaction sequence number for cursor fetches, even if subsequent
writes happen and change that sequence number in the user's transaction.

Release note: None

jordanlewis committed:
sql: ban mutations from cursors; eagerly return plan errors

Cursors are not allowed to run mutations via common-table expressions.
Ban them by running the optimizer on the statement up front.

Also, this allows us to return any plan errors eagerly on DECLARE,
rather than waiting for FETCH.

Release note: None

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74006)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jordanlewis, @otan, @rafiss, and @yuzefovich)

___
*[pkg/sql/sql_cursor.go, line 132 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-MrY45xZ74mjHwSXGu21:-MriTitH-O7U4txZbLam:b-abda1g) ([raw file](https://github.com/cockroachdb/cockroach/blob/b21c6703a2078f37f07ee4ecc5fe2b6380e0074e/pkg/sql/sql_cursor.go#L132)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Relative fetch can also be backward, should we check that `f.offset` is not negative?
</blockquote></details>

This is handled elsewhere, but I've added a test for it.
___
*[pkg/sql/sql_cursor.go, line 175 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-MrY2slW7Iz032GF9Tc5:-MriUAn94mWCDY44i7m7:b-gp4a2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/b21c6703a2078f37f07ee4ecc5fe2b6380e0074e/pkg/sql/sql_cursor.go#L175)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: this probably could be unexported.
</blockquote></details>

Well, it shadows the .Next on the embedded InternalRows, so I think it's better to keep exported. Added comment.
___
*[pkg/sql/parser/sql.y, line 5092 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-MrULCly4eWohgHPq354:-MriURF3412_k1e9fc_b:b-h207fl) ([raw file](https://github.com/cockroachdb/cockroach/blob/10edf738171e7d94bff99a94b14941a15a1dc4f9/pkg/sql/parser/sql.y#L5092)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

PG docs say that
```
The key words ASENSITIVE, BINARY, INSENSITIVE, and SCROLL can appear in any order.
```
it looks like we currently don't support that, right? Is the only way to support that is to have all permutations defined explicitly?
</blockquote></details>

Yes, that's true... The postgres grammar does this in a different way, I will try to do this as well.
___
*[pkg/sql/sem/tree/cursor.go, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-MrY0CyX09BmlKoqcBHb:-MriVSIj0fP6aiMtqN00:bc1hsld) ([raw file](https://github.com/cockroachdb/cockroach/blob/10edf738171e7d94bff99a94b14941a15a1dc4f9/pkg/sql/sem/tree/cursor.go#L42)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Does it not matter to distinguish between unspecified hold and explicit "without hold"?
</blockquote></details>

No it doesn't appear to.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
Update: the "insensitivity" property is a bit more tricky to produce than I originally thought. The reason is that the internal executor that we create to run the cursor just uses the user transaction as normal - it doesn't "capture" any transaction state. This means that any subsequent writes in the transaction actually do edit the read sequence number of the transaction that the internal executor is using, and thus will make the cursor "sensitive" - subsequent writes in the transaction will be visible by a cursor that was declared earlier.

It seems like there are 2 ways to handle this, after discussions with @nvanbenschoten.

One is to use a "leaf transaction" with a fixed read sequence number, like we do in DistSQL. But leaf transactions have some limitations: you can't spawn more leaf transactions from them, which means that you couldn't run a DistSQL query in a cursor. And, you also can't set a deadline on a leaf transaction, which is used during query setup. So, this strategy would require some extra work.

The other is to update the transaction's read sequence number "on the fly". This is only safe if the cursor's transaction is not used concurrently with the user's transaction. But, since the cursor's transaction is running in an internal executor, which is streaming and "frozen" after each call to `.Next`, we can in fact guarantee this.

So, what this PR does is update the user's transaction's read sequence number when running `FETCH` on the cursor, and restore the sequence number when the `FETCH` is complete.

This appears to work, but it could possibly be broken, I'm not too sure.

andreimatei commented:
> But leaf transactions have some limitations: you can't spawn more leaf transactions from them

Why not? I think leaf txns are fungible; you can take the `LeafTxnInputState` and create as many as you want. No?

> And, you also can't set a deadline on a leaf transaction

Why not? It's kinda the other way around - a `LeafTxn` cannot have its timestamp advanced, so the notion of a deadline is not exactly necessary. But mechanically I think they can have a deadline.

knz commented:
> So, what this PR does is update the user's transaction's read sequence number when running FETCH on the cursor, and restore the sequence number when the FETCH is complete.

The issue with that is if some operation inside the plan moves the read seqnum forward during a FETCH operation, then that cursor will be able to observe writes that have logically been performed "after" it was declared.

You need a way to freeze the read seqnum, not just rewind it each time.

a LeafTxn might help with that.

knz reviewed (commented):
One thing that would help with this development would be to enumerate an exhaustive set of basic test cases, for all the possible interleavings of a cursor with 1) a concurrent txn (including both read-only and read-write) and 2) savepoints in the main txn.

The test cases would then be used to both exercise the proposed solution and serve as basis to document special considerations in usage, if any.

I see there are some test cases in this PR but it's not exactly exhaustive yet.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-:-Mrlr26Y59U6hiZi90iH:by2fp8g)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jordanlewis, @otan, @rafiss, and @yuzefovich)



<!-- Sent from Reviewable.io -->

knz reviewed (commented):
(sorry in my previous comment I meant interleavings of a cursor with further statements in the main txn)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-:-Mrlrdlc5UNgYZ97HRX0:b-hqyd6s)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jordanlewis, @otan, @rafiss, and @yuzefovich)



<!-- Sent from Reviewable.io -->

jordanlewis commented:
> The issue with that is if some operation inside the plan moves the read seqnum forward during a FETCH operation, then that cursor will be able to observe writes that have logically been performed "after" it was declared.

What operations inside of a plan move a read seqnum forward beside mutations? 

> Why not? I think leaf txns are fungible; you can take the LeafTxnInputState and create as many as you want. No?

It's currently banned with an assertion in `GetLeafTxnInputState`, and I don't know enough about why it's banned to try to circumvent it:

```
	if txn.typ != RootTxn {
		panic(errors.WithContextTags(errors.AssertionFailedf("GetLeafTxnInputState() called on leaf txn"), ctx))
	}
```

> Why not? It's kinda the other way around - a LeafTxn cannot have its timestamp advanced, so the notion of a deadline is not exactly necessary. But mechanically I think they can have a deadline.

It's currently banned with an assertion in `Txn.UpdateDeadline` as well, and I don't know enough about why it's banned to try to circumvent it:

```
func (txn *Txn) UpdateDeadline(ctx context.Context, deadline hlc.Timestamp) error {
	if txn.typ != RootTxn {
		panic(errors.WithContextTags(errors.AssertionFailedf("UpdateDeadline() called on leaf txn"), ctx))
	}
```

rafiss commented:
I hope to give this a closer review soon, but I coincidentally ran into this part of the Postgres docs while working on something else:
> Named portals [from the extended protocol] can also be created and accessed at the SQL command level, using DECLARE CURSOR and FETCH.

https://www.postgresql.org/docs/14/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY

So I wonder if the implementation for DECLARE/FETCH could actually share code with how we've implemented portals, and if that would ultimately be "more correct" anyway.

jordanlewis commented:
> So I wonder if the implementation for DECLARE/FETCH could actually share code with how we've implemented portals, and if that would ultimately be "more correct" anyway.

Trying to do this has the same challenges as the effort to support multiple active result sets (#40195). I'm guessing that most of the time people aren't really using `DECLARE/FETCH` to interoperate with the namespace of portals in the extended protocol. It would definitely be cool to have these all share an implementation but it's a big lift due to all of the shared state in the executor and planner.

knz commented:
> What operations inside of a plan move a read seqnum forward beside mutations?

- A lot of of DDL
- Job-creating operations (IIRC)
- certain privilege changes

One thing I'm annoyed (worried?) at is that we've never designed the seqnum machinery (and the corresponding `.Next()` call) with the need to create a strict inventory of the statements that use it vs those that don't.  So now we're not even sure whether they overlap with postgres.

otan reviewed (commented):
this looks really neat! appreciate the ball rollin'.

there seems to be a lot of reviews on this already; let me know if you think it's worth my input on implementation (from my reading it seems like it may change around a bit still! from the KV leaf txn stuff, but may be wrong).

in terms of functionality - how easy is it to support `BINARY CURSOR`? i think i've encountered some things that need it.

otan commented on pkg/sql/parser/sql.y:
```diff
@@ -5046,14 +5065,204 @@ show_stmt:
+// %SeeAlso: CLOSE, FETCH
 declare_cursor_stmt:
-	DECLARE { return unimplementedWithIssue(sqllex, 41412) }
+	DECLARE cursor_name opt_binary opt_sensitivity opt_scroll CURSOR opt_hold FOR select_stmt
```
i think these cursor_options can be in any order in PG, can we match?

https://github.com/postgres/postgres/blob/8a2e323f208557acbfdc911516e84ec017f5a6ca/src/backend/parser/gram.y#L11437

jordanlewis replied:
I'm not sure the best way to do this.

otan replied:
can we do something like

```
cursor_option:
    BINARY { BinaryCursorOption } 
|   ...

opt_cursor_options:
   cursor_options { return $1 }
| /* blank */ { return []CursorOption{} }

cursor_options:
    cursor_option { []CursorOption{...} }
    cursor_options ',' cursor_option { return append($1, $3) }
```

And have `CursorOption` implement an interface `ApplyCursorOption` which applies some sort of method onto the `Cursor` AST node?

you can leave it as a TODO, not too fussed

jordanlewis replied:
I left a TODO 😅

otan commented on pkg/sql/parser/sql.y:
```diff
@@ -5046,14 +5065,204 @@ show_stmt:
+      Name: tree.Name($3),
+      FetchType: tree.FetchLast,
+    }
+  }
```
i think we're missing `SignedIConst`: https://github.com/postgres/postgres/blob/8a2e323f208557acbfdc911516e84ec017f5a6ca/src/backend/parser/gram.y#L6880-L6887

jordanlewis replied:
This is allowed by this rule:

```
| opt_forward_backward signed_iconst64 opt_from_or_in cursor_name
```

otan commented on pkg/sql/parser/sql.y:
```diff
@@ -5046,14 +5065,204 @@ show_stmt:
+  FETCH fetch_specifier
+  {
+    $$.val = $2.fetchCursor()
+  }
```
did you want to add boilerplate for `MOVE` as well?
looks like it's implemented as the same `FETCH` node in pg: https://github.com/postgres/postgres/blob/8a2e323f208557acbfdc911516e84ec017f5a6ca/src/backend/parser/gram.y#L6802-L6813

jordanlewis replied:
Maybe this could come as another PR? This is already quite a beefy one.

otan commented on pkg/sql/sql_cursor.go:
```diff
@@ -13,24 +13,169 @@ package sql
+	if s.Scroll == tree.Scroll {
+		return nil, unimplemented.NewWithIssue(41412, "DECLARE SCROLL CURSOR")
+	}
+	if s.Sensitivity != tree.UnspecifiedSensitivity {
```
i think we can just let this "pass through":

```
ASENSITIVE
INSENSITIVE
Cursor sensitivity determines whether changes to the data underlying the cursor, done in the same transaction, after the cursor has been declared, are visible in the cursor. INSENSITIVE means they are not visible, ASENSITIVE means the behavior is implementation-dependent. A third behavior, SENSITIVE, meaning that such changes are visible in the cursor, is not available in PostgreSQL. In PostgreSQL, all cursors are insensitive; so these key words have no effect and are only accepted for compatibility with the SQL standard.

Specifying INSENSITIVE together with FOR UPDATE or FOR SHARE is an error.
```

https://www.postgresql.org/docs/current/sql-declare.html

jordanlewis replied:
Done.

otan reviewed (commented):
i lied, i took a quick squiz!

jordanlewis commented:
> in terms of functionality - how easy is it to support BINARY CURSOR? i think i've encountered some things that need it.

I think this should be fairly straightforward, but I'm not sure. I think the way it would work is that you'd replace all of the column types of a query with bytes, and add code that runs the pgwire binary formatter at the end of the query.

jordanlewis commented:
> > What operations inside of a plan move a read seqnum forward beside mutations?
> A lot of of DDL
> Job-creating operations (IIRC)
> certain privilege changes

DDLs (and, in general, non-preparable statements, as well as mutations) are not allowed in cursor statements. So I think we are good to go here.

jordanlewis commented:
> My first instinct is that this is going to have bad interactions with schema changes in the transaction. That is to say, it won't see them. That's pretty gross. I wonder what happens when you create cursors over various pg_catalog things and then modify them while holding the cursor open. I'm not even sure what the right thing is.


@ajwerner here's what Postgres does, so I guess we can do this kind of thing too somehow.

```
jordan=> begin;
BEGIN
jordan=> declare foo no scroll cursor for select a::text|| 'hi'::text from a;
DECLARE CURSOR
jordan=> alter table a drop column b;
ERROR:  cannot ALTER TABLE "a" because it is being used by active queries in this session
jordan=>
```

ajwerner commented:
> @ajwerner here's what Postgres does, so I guess we can do this kind of thing too somehow.

That's a very satisfying behavior. I like it when postgres does sane things. I would think we can leverage the set of dependencies which the optimizer tracks for `IsStale`: https://github.com/cockroachdb/cockroach/blob/ad27839166526a92444bf0f719cf9ec5f62c2f7f/pkg/sql/opt/metadata.go#L273-L284

jordanlewis commented:
Is there already an appropriate place in the beginning phases of schema changes in which I could insert checks against all active cursors?

ajwerner commented:
> Is there already an appropriate place in the beginning phases of schema changes in which I could insert checks against all active cursors?

I think here's a pretty good spot:
https://github.com/cockroachdb/cockroach/blob/ad27839166526a92444bf0f719cf9ec5f62c2f7f/pkg/sql/opaque.go#L50

jordanlewis commented:
Yes, but we haven't parsed apart the statement yet. Until the new schema changer lands, it's inconvenient - I think we'd have to do it in 2 places.

Hmm, my intuition is telling me that banning schema changes from transactions that have open cursors is going to affect nobody. How would you feel if I did that for now, with an issue and TODO for post-new-schema-changer?

ajwerner commented:
> Yes, but we haven't parsed apart the statement yet. Until the new schema changer lands, it's inconvenient - I think we'd have to do it in a few places (i.e. in new schema changer after parsing, and once per statement in old schema changer(?)).
> 
> Hmm, my intuition is telling me that banning schema changes from transactions that have _any_ open cursors is going to affect nobody. How would you feel if I did that for now, with an issue and TODO for post-new-schema-changer?

I'm pleased with that.

jordanlewis committed:
sql: bar schema change in txns w/ open sql cursors

This commit prevents schema changes from being initiated in open
transactions that have open cursors inside. This is to prevent issues
with cursors that remain open on a table that might get modified by a
schema change.

In Postgres, only schema changes that modify objects used by open
cursors are barred. For now, we just bar all schema changes because we
don't expect that this will affect many users, and it's a bit annoying
to do this precisely until the new schema changer lands.

Release note: None

jordanlewis commented:
Added the "no schema changes" limitation in another commit.

This is ready for another review! @otan any further review from you is definitely appreciated.

I'm at least somewhat convinced that the txn seqnum implementation I've created here makes sense. Using a LeafTxn seems to be impractical without a fairly large refactor of LeafTxn due to limitations I've touched on above.

@knz, as far as the test case enumeration you've proposed, I've not done it yet because I don't yet see what an interesting condition would be besides the simple cases of modify-after-cursor-creation and modify-before-cursor-creation-in-same-txn which I've already exercised. Concurrent txn isn't interesting because it's already exercised exactly the same by any of our other concurrent txn testing.

knz commented:
@nvanbenschoten can you help ascertain that the seqnum dance performed here makes sense?

otan reviewed (approved):
looks good to me, i'm unsure how to sign off on the seq num stuff though.

nvb reviewed (commented):
The seqnum dance here makes sense to me. I think it would be worthwhile to add some kv-level testing for it, like we have for stepping modes in `pkg/kv/kvclient/kvcoord/txn_interceptor_seq_num_allocator_test.go`.

> Reviewed 12 of 16 files at r1, 2 of 15 files at r17, 15 of 15 files at r23, 6 of 6 files at r24, 4 of 4 files at r25, 7 of 7 files at r26, 6 of 6 files at r27, 2 of 2 files at r28, 3 of 3 files at r29, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-:-Mv5oi2C1ki82XLH-4Ds:bf10izn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jordanlewis, @otan, @rafiss, and @yuzefovich)

___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender.go, line 1309 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-Mv5nEYB0-4LAMkHRjov:-Mv5nEYB0-4LAMkHRjow:bcjv2rs) ([raw file](https://github.com/cockroachdb/cockroach/blob/958c7ea6c54407232e9929a9f549619df0f23d62/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L1309)):*
> ```Go
> 	tc.mu.Lock()
> 	defer tc.mu.Unlock()
> 	tc.interceptorAlloc.txnSeqNumAllocator.readSeq = seq
> ```

Should we add any validation here, like that this is >= 0 and <= writeSeq?
___
*[pkg/sql/sql_cursor.go, line 52 at r25](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-Mr9t_6d4c0J9hblbQPg-r25-52:-Mv5OvZt1iW8Pu2VaBRe:bxa5vxz) ([raw file](https://github.com/cockroachdb/cockroach/blob/60acbc3cf6cd34980882c6825f1ce618f50fc227/pkg/sql/sql_cursor.go#L52)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

missing a comment line by the looks of it
</blockquote></details>

Yes, this looks off.
___
*[pkg/sql/sql_cursor.go, line 180 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-Mv5p3ty3Q-zjE6UCQNB:-Mv5p3ty3Q-zjE6UCQNC:b-xbkn2r) ([raw file](https://github.com/cockroachdb/cockroach/blob/958c7ea6c54407232e9929a9f549619df0f23d62/pkg/sql/sql_cursor.go#L180)):*
> ```Go
> 	// running FETCH on a CURSOR does not close it.
> 
> 	// Reset the trasnaction's read sequence number to what it was before the
> ```

s/trasnaction/transaction/
___
*[pkg/sql/parser/sql.y, line 5122 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-Mv5GU2b05RpSvU22S0X:-Mv5GU2b05RpSvU22S0Y:b-hh2csj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e96402fda8a4f4ec57185b322fcddd07927b8e2a/pkg/sql/parser/sql.y#L5122)):*
> ```Yacc
>     $$.val = tree.Insensitive
>   }
> | ASENSITIVE
> ```

Do we like that we won't even parse SENSITIVE? Would it be better to include that in the grammar but then throw a descriptive error?
___
*[pkg/sql/sem/tree/cursor.go, line 51 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-Mv5GBxY72JQMOVH1Mp3:-Mv5GBxY72JQMOVH1Mp4:b8n8ldu) ([raw file](https://github.com/cockroachdb/cockroach/blob/e96402fda8a4f4ec57185b322fcddd07927b8e2a/pkg/sql/sem/tree/cursor.go#L51)):*
> ```Go
> // CursorScrollOption represents the scroll option, if one was given, for a
> // DECLARE statement.
> type CursorScrollOption int
> ```

nit: should we make these enum types `int8`? That will allow `DeclareCursor` to pack better.
___
*[pkg/sql/logictest/testdata/logic_test/cursor, line 99 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-Mv5o1di2LKfexmafqwf:-Mv5o1di2LKfexmafqwg:b-6kja66) ([raw file](https://github.com/cockroachdb/cockroach/blob/958c7ea6c54407232e9929a9f549619df0f23d62/pkg/sql/logictest/testdata/logic_test/cursor#L99)):*
> ```
> 
> query IT
> FETCH RELATIVE 12 foo
> ```

Should we perform another INSERT after this first fetch as well, to make sure the cursor is insensitive to that as well?
___
*[pkg/sql/logictest/testdata/logic_test/cursor, line 99 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-Mv5oVN-60GRuf3R2uTm:-Mv5oVN-60GRuf3R2uTn:bdnfblm) ([raw file](https://github.com/cockroachdb/cockroach/blob/958c7ea6c54407232e9929a9f549619df0f23d62/pkg/sql/logictest/testdata/logic_test/cursor#L99)):*
> ```
> 
> query IT
> FETCH RELATIVE 12 foo
> ```

Should we perform a read-only operation outside of the cursor to demonstrate that it _is_ sensitive to these INSERTs and to demonstrate the cursor hasn't corrupted the read seq num state?


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
TFTR!

I added a test. I don't think it's that great. Basically it just makes sure that the txnSeqNumAllocator doesn't choke when you edit the read seqnum and send more commands. I'm just not sure if this is testing much. Did you have something else in mind as well?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-:-MwDjO_t6FSnz2GmPIvc:bm5i6jy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jordanlewis, @nvanbenschoten, @otan, @rafiss, and @yuzefovich)

___
*[pkg/kv/kvclient/kvcoord/txn_coord_sender.go, line 1309 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-Mv5nEYB0-4LAMkHRjov:-MwDZnKC4DZHbya7-fwi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/958c7ea6c54407232e9929a9f549619df0f23d62/pkg/kv/kvclient/kvcoord/txn_coord_sender.go#L1309)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we add any validation here, like that this is >= 0 and <= writeSeq?
</blockquote></details>

Done.
___
*[pkg/sql/sql_cursor.go, line 52 at r25](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-Mr9t_6d4c0J9hblbQPg-r25-52:-MwD_kDUA9upWKMY2sWU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/60acbc3cf6cd34980882c6825f1ce618f50fc227/pkg/sql/sql_cursor.go#L52)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Yes, this looks off.
</blockquote></details>

Done.
___
*[pkg/sql/parser/sql.y, line 5122 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-Mv5GU2b05RpSvU22S0X:-MwDa2L68p1pw6fzDsZ3:b-hnufy) ([raw file](https://github.com/cockroachdb/cockroach/blob/e96402fda8a4f4ec57185b322fcddd07927b8e2a/pkg/sql/parser/sql.y#L5122)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we like that we won't even parse SENSITIVE? Would it be better to include that in the grammar but then throw a descriptive error?
</blockquote></details>

SENSITIVE isn't accepted by pg either so this would be a custom extension. I feel comfortable omitting it.
___
*[pkg/sql/sem/tree/cursor.go, line 51 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-Mv5GBxY72JQMOVH1Mp3:-MwDaHKw5yOTDidJaGvh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e96402fda8a4f4ec57185b322fcddd07927b8e2a/pkg/sql/sem/tree/cursor.go#L51)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: should we make these enum types `int8`? That will allow `DeclareCursor` to pack better.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/cursor, line 25 at r24](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-MrA2pQPM2YZS24rT71T-r24-25:-MwDc9K49EoMoxzl4hqr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5d6333cc208bb718e6ce54a5b83b8fcdc560f74/pkg/sql/logictest/testdata/logic_test/cursor#L25)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

can you add a test for a cursor which errors, e.g. selects a column which doesn't exist in a table?
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/cursor, line 99 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-Mv5o1di2LKfexmafqwf:-MwDg-HDCQnxoA8HdTho:b-fznogt) ([raw file](https://github.com/cockroachdb/cockroach/blob/958c7ea6c54407232e9929a9f549619df0f23d62/pkg/sql/logictest/testdata/logic_test/cursor#L99)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we perform another INSERT after this first fetch as well, to make sure the cursor is insensitive to that as well?
</blockquote></details>

Good idea, done.
___
*[pkg/sql/logictest/testdata/logic_test/cursor, line 99 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/74006#-Mv5oVN-60GRuf3R2uTm:-MwDg-yPEIs-hDPNknt5:b-fznogt) ([raw file](https://github.com/cockroachdb/cockroach/blob/958c7ea6c54407232e9929a9f549619df0f23d62/pkg/sql/logictest/testdata/logic_test/cursor#L99)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we perform a read-only operation outside of the cursor to demonstrate that it _is_ sensitive to these INSERTs and to demonstrate the cursor hasn't corrupted the read seq num state?
</blockquote></details>

Good idea, done.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
I'm going to go ahead and merge this as I've resolved all outstanding comments, see no show-stopping issues with the KV interaction thanks to @nvanbenschoten's review, and have approval from @otan. Thanks for the reviews, all!

bors r+

rafiss commented:
Looks like this caused a panic in https://github.com/cockroachdb/cockroach/issues/76883#issuecomment-1048175849

It seems related to using a DECLARE statement in the extended protocol. One thing I noticed is this condition: https://github.com/cockroachdb/cockroach/blob/96d102af5e937677298184f848ac9636f95985ec/pkg/sql/sql_cursor.go#L33

It probably is not what we want to be checking when in the extended protocol. `p.autocommit` will be false for implicit transactions in the extended protocol.
